File: flutter_app/lib/services/api_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

class ApiService {
  static final ApiService _instance = ApiService._internal();
  factory ApiService() => _instance;

  late Dio _dio;
  String? _token;
  static GlobalKey<NavigatorState>? navigatorKey;

  ApiService._internal() {
    _dio = Dio(BaseOptions(
      baseUrl: 'http://120.27.115.89:8080/api',
      connectTimeout: const Duration(seconds: 10),
      receiveTimeout: const Duration(seconds: 10),
      headers: {'Content-Type': 'application/json'},
    ));

    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        if (_token != null) {
          options.headers['Authorization'] = 'Bearer $_token';
        }
        return handler.next(options);
      },
      onError: (error, handler) async {
        if (error.response?.statusCode == 401) {
          await clearToken();
          _navigateToLogin();
        }
        return handler.next(error);
      },
    ));
  }

  Dio get dio => _dio;

  Future<void> setToken(String token) async {
    _token = token;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('jwt_token', token);
  }

  Future<void> loadToken() async {
    final prefs = await SharedPreferences.getInstance();
    _token = prefs.getString('jwt_token');
  }

  Future<void> clearToken() async {
    _token = null;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('jwt_token');
  }

  bool get hasToken => _token != null;
  String? get token => _token;

  void _navigateToLogin() {
    if (navigatorKey?.currentState != null) {
      navigatorKey!.currentState!.pushNamedAndRemoveUntil('/login', (route) => false);
    }
  }

  // Auth
  Future<Response> register(String username, String phoneNumber, String password) async {
    return _dio.post('/auth/register', data: {
      'username': username,
      'phone_number': phoneNumber,
      'password': password,
    });
  }

  Future<Response> login(String account, String password) async {
    return _dio.post('/auth/login', data: {
      'account': account,  // 用户名或手机号
      'password': password,
    });
  }

  // 验证token是否有效
  Future<bool> verifyToken() async {
    if (_token == null) return false;
    try {
      await _dio.get('/auth/verify');
      return true;
    } catch (e) {
      await clearToken();
      return false;
    }
  }

  // 获取用户信息
  Future<Response> getProfile() async {
    return _dio.get('/user/profile');
  }

  // 更新用户信息
  Future<Response> updateProfile({String? nickname, String? avatar}) async {
    final data = <String, dynamic>{};
    if (nickname != null) data['nickname'] = nickname;
    if (avatar != null) data['avatar'] = avatar;
    return _dio.put('/user/profile', data: data);
  }

  // 修改密码
  Future<Response> changePassword(String oldPassword, String newPassword) async {
    return _dio.put('/user/password', data: {
      'old_password': oldPassword,
      'new_password': newPassword,
    });
  }

  // Plans
  Future<Response> getPlans({String? date}) async {
    return _dio.get('/plans', queryParameters: date != null ? {'date': date} : null);
  }

  Future<Response> createPlan(String content, String executionDate) async {
    return _dio.post('/plans', data: {
      'content': content,
      'execution_date': executionDate,
    });
  }

  Future<Response> updatePlan(int id, {String? content, String? status}) async {
    final data = <String, dynamic>{};
    if (content != null) data['content'] = content;
    if (status != null) data['status'] = status;
    return _dio.put('/plans/$id', data: data);
  }

  Future<Response> deletePlan(int id) async {
    return _dio.delete('/plans/$id');
  }

  // Expenses
  Future<Response> getExpenses() async {
    return _dio.get('/expenses');
  }

  Future<Response> createExpense(double amount, String category, String? note) async {
    return _dio.post('/expenses', data: {
      'amount': amount,
      'category': category,
      'note': note ?? '',
    });
  }

  Future<Response> updateExpense(int id, {double? amount, String? category, String? note}) async {
    final data = <String, dynamic>{};
    if (amount != null) data['amount'] = amount;
    if (category != null) data['category'] = category;
    if (note != null) data['note'] = note;
    return _dio.put('/expenses/$id', data: data);
  }

  Future<Response> deleteExpense(int id) async {
    return _dio.delete('/expenses/$id');
  }

  // Reminders
  Future<Response> getReminders() async {
    return _dio.get('/reminders');
  }

  Future<Response> createReminder(String type, String scheduledTime, String? content) async {
    return _dio.post('/reminders', data: {
      'reminder_type': type,
      'scheduled_time': scheduledTime,
      'content': content ?? '',
    });
  }

  Future<Response> updateReminder(int id, {String? scheduledTime, String? content, bool? isEnabled}) async {
    final data = <String, dynamic>{};
    if (scheduledTime != null) data['scheduled_time'] = scheduledTime;
    if (content != null) data['content'] = content;
    if (isEnabled != null) data['is_enabled'] = isEnabled;
    return _dio.put('/reminders/$id', data: data);
  }

  Future<Response> deleteReminder(int id) async {
    return _dio.delete('/reminders/$id');
  }
}

File: swift_app/SwiftApp/Services/APIService.swift
// APIService.swift
// Network service for API communication

import Foundation

enum APIError: Error, LocalizedError {
    case invalidURL
    case invalidResponse
    case unauthorized
    case serverError(String)
    case decodingError
    case networkError(Error)
    
    var errorDescription: String? {
        switch self {
        case .invalidURL: return "无效的URL"
        case .invalidResponse: return "服务器响应异常"
        case .unauthorized: return "登录已过期，请重新登录"
        case .serverError(let msg): return msg
        case .decodingError: return "数据解析失败"
        case .networkError(let error): return error.localizedDescription
        }
    }
}

@MainActor
class APIService: ObservableObject {
    static let shared = APIService()
    
    private let baseURL = "http://120.27.115.89:8080/api"
    private var token: String?
    
    private init() {
        loadToken()
    }
    
    func setToken(_ token: String) {
        self.token = token
        UserDefaults.standard.set(token, forKey: "jwt_token")
    }
    
    func loadToken() {
        token = UserDefaults.standard.string(forKey: "jwt_token")
    }
    
    func clearToken() {
        token = nil
        UserDefaults.standard.removeObject(forKey: "jwt_token")
    }
    
    var hasToken: Bool { token != nil }
    
    // MARK: - Request Helper
    
    private func request<T: Codable>(
        endpoint: String,
        method: String = "GET",
        body: [String: Any]? = nil
    ) async throws -> T {
        guard let url = URL(string: baseURL + endpoint) else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let token = token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        if let body = body {
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        }
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw APIError.invalidResponse
            }
            
            if httpResponse.statusCode == 401 {
                clearToken()
                throw APIError.unauthorized
            }
            
            if httpResponse.statusCode >= 400 {
                if let errorMsg = try? JSONDecoder().decode([String: String].self, from: data),
                   let msg = errorMsg["error"] ?? errorMsg["message"] {
                    throw APIError.serverError(msg)
                }
                throw APIError.serverError("请求失败")
            }
            
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
        } catch let error as APIError {
            throw error
        } catch let error as DecodingError {
            print("Decoding error: \(error)")
            throw APIError.decodingError
        } catch {
            throw APIError.networkError(error)
        }
    }
    
    // MARK: - Auth
    
    func login(account: String, password: String) async throws -> AuthResponse {
        struct LoginRequest: Codable { let account, password: String }
        
        guard let url = URL(string: baseURL + "/auth/login") else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(LoginRequest(account: account, password: password))
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }
        
        if httpResponse.statusCode >= 400 {
            if let errorMsg = try? JSONDecoder().decode([String: String].self, from: data),
               let msg = errorMsg["error"] ?? errorMsg["message"] {
                throw APIError.serverError(msg)
            }
            throw APIError.serverError("登录失败")
        }
        
        let authResponse = try JSONDecoder().decode(AuthResponse.self, from: data)
        setToken(authResponse.token)
        return authResponse
    }
    
    func register(username: String, phoneNumber: String, password: String) async throws -> AuthResponse {
        guard let url = URL(string: baseURL + "/auth/register") else {
            throw APIError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: String] = [
            "username": username,
            "phone_number": phoneNumber,
            "password": password
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }
        
        if httpResponse.statusCode >= 400 {
            if let errorMsg = try? JSONDecoder().decode([String: String].self, from: data),
               let msg = errorMsg["error"] ?? errorMsg["message"] {
                throw APIError.serverError(msg)
            }
            throw APIError.serverError("注册失败")
        }
        
        let authResponse = try JSONDecoder().decode(AuthResponse.self, from: data)
        setToken(authResponse.token)
        return authResponse
    }
    
    func verifyToken() async throws -> Bool {
        guard token != nil else { return false }
        do {
            let _: [String: String] = try await request(endpoint: "/auth/verify")
            return true
        } catch {
            clearToken()
            return false
        }
    }
    
    func getProfile() async throws -> User {
        return try await request(endpoint: "/user/profile")
    }
    
    func updateProfile(nickname: String?, avatar: String?) async throws -> User {
        var body: [String: String] = [:]
        if let nickname = nickname { body["nickname"] = nickname }
        if let avatar = avatar { body["avatar"] = avatar }
        return try await request(endpoint: "/user/profile", method: "PUT", body: body)
    }
    
    func changePassword(oldPassword: String, newPassword: String) async throws {
        let _: [String: String] = try await request(
            endpoint: "/user/password",
            method: "PUT",
            body: ["old_password": oldPassword, "new_password": newPassword]
        )
    }
    
    // MARK: - Plans
    
    func getPlans(date: String? = nil) async throws -> [Plan] {
        var endpoint = "/plans"
        if let date = date { endpoint += "?date=\(date)" }
        return try await request(endpoint: endpoint)
    }
    
    func createPlan(content: String, executionDate: String) async throws -> Plan {
        return try await request(
            endpoint: "/plans",
            method: "POST",
            body: ["content": content, "execution_date": executionDate]
        )
    }
    
    func updatePlan(id: Int, content: String? = nil, status: String? = nil) async throws -> Plan {
        var body: [String: String] = [:]
        if let content = content { body["content"] = content }
        if let status = status { body["status"] = status }
        return try await request(endpoint: "/plans/\(id)", method: "PUT", body: body)
    }
    
    func deletePlan(id: Int) async throws {
        let _: [String: String] = try await request(endpoint: "/plans/\(id)", method: "DELETE")
    }
    
    // MARK: - Expenses
    
    func getExpenses() async throws -> [Expense] {
        return try await request(endpoint: "/expenses")
    }
    
    func createExpense(amount: Double, category: String, note: String?) async throws -> Expense {
        return try await request(
            endpoint: "/expenses",
            method: "POST",
            body: ["amount": amount, "category": category, "note": note ?? ""]
        )
    }
    
    func deleteExpense(id: Int) async throws {
        let _: [String: String] = try await request(endpoint: "/expenses/\(id)", method: "DELETE")
    }
}
